### DISCUSSION TITLE
Using ConfigureAwait in "await using" declaration

### DISCUSSION THREAD

**Discussion by cincuranet** [dotnet/csharplang#2661] [Q&A] [answered]
**Version Used**: 
3.0.100-preview3-010431

**Steps to Reproduce**:

When running the code below, the compiler says `error CS1061: 'ConfiguredAsyncDisposable' does not contain a definition for 'Dummy' and no accessible extension method 'Dummy' accepting a first argument of type 'ConfiguredAsyncDisposable' could be found (are you missing a using directive or an assembly reference?)` which makes absolute sense.

```csharp
namespace test
{
	class AwaitUsingConfigureAwaitTest: IAsyncDisposable
	{
		public async ValueTask DisposeAsync()
		{
			await Task.CompletedTask;	
		}

		public async Task FooBar()
		{
			await using (var test = new AwaitUsingConfigureAwaitTest().ConfigureAwait(false))
			{
				test.Dummy();
			}
		}

		public void Dummy() { }
	}
}
```

That means the code has to be written like this.

```csharp
public async Task FooBar()
{
	var test = new AwaitUsingConfigureAwaitTest();
	await using (test.ConfigureAwait(false))
	{
		test.Dummy();
	}
}
```

Not as succinct as before. 

Is this the final state of this feature or is there some work still pending around this area?

**Comment by jcouv** [comment]
Tagging @stephentoub  as well

**Comment by agocke** [comment, marked as answer]
This is the final state of the feature for C# 8.

**Comment by DenSmoke** [comment]
```C#
var connection = new SqlConnection(Options.ConnectionString);
await using (connection.ConfigureAwait(false))
{
    var command = new SqlCommand(Query, connection) { CommandTimeout = Options.CommandTimeout };
    await using (command.ConfigureAwait(false))
    {
        await connection.OpenAsync().ConfigureAwait(false);
        var sqlDependency = new SqlDependency(command);
        sqlDependency.OnChange += SqlDependencyOnChange;
        CurrentSqlDependency = sqlDependency;
        await using var reader = (await command.ExecuteReaderAsync().ConfigureAwait(false)).ConfigureAwait(false);
    }
}
```
Instead of this:
```C#
await using var sc = new SqlConnection(Options.ConnectionString);
await using var command = new SqlCommand(Query, sc) { CommandTimeout = Options.CommandTimeout };
await sc.OpenAsync().ConfigureAwait(false);
var sqlDependency = new SqlDependency(command);
sqlDependency.OnChange += SqlDependencyOnChange;
CurrentSqlDependency = sqlDependency;
await using var reader = await command.ExecuteReaderAsync().ConfigureAwait(false);
```
I think it needs to be reviewed. I almost don't know situations when we need to continue on captured context. Even in UI I just use something like Dispatcher.Invoke or SynchronizationContext.Post and don't worry about context.

**Comment by qrli** [comment]
The `ConfigureAwait(false)` thing again and it is becoming more messy.
Instead of continuing this, I'd propose to have a standard way to leave the current synchronization context. It could be an API in corefx, not necessarily a language feature.

As some examples from discussions here before:
```cs
async Task Foo() {
  await AsyncHelper.LeaveSynchronizationContext();
  ... // effectively all awaits here behave like ConfigureAwait(false)'ed
}
```
Or
```cs
async Task Foo() {
  using (AsyncHelper.NoSynchronizationContext()) {
    ... // effectively all awaits here behave like ConfigureAwait(false)'ed
  }
}
```
Or with some compiler magic to make it nicer:
```cs
[ConfigureAwait(false)] // or maybe name it [NoSynchronizationContext]
async Task Foo() {
  ... // effectively all awaits here behave like ConfigureAwait(false)'ed
}
```
This can also relief the complains on `ConfigureAwait(false)`.

**Reply by martingbrown** [reply]
I had the idea that we could have a modifier for the await keyword that would be both easy to type and do the .ConfigureAwait(false) thing. For example adding a ~ to it. So `await~ SomeMethod();` would be expanded by the compiler to `await SomeMethod().ConfigureAwait(false);`. The same could be used in the `await using` case with `await~ using`.

**Reply by colejohnson66** [reply]
Please no keywords. They are cryptic and not expandable, especially now that [other `ConfigureAwait` options are available](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.configureawaitoptions?view=net-8.0).

**Comment by adams85** [comment]
You can also workaround this by abusing out parameter variable declarations:

```csharp
public static class AsyncDisposableHelpers
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static IAsyncDisposable AsAsyncDisposable<T>(this T disposable, out T disposableOut) where T : notnull, IAsyncDisposable =>
        disposableOut = disposable;
}
```

then you can write

```csharp
await using (new AwaitUsingConfigureAwaitTest().AsAsyncDisposable(out var test).ConfigureAwait(false))
{
    test.Dummy();
}
```

**Comment by krisztiankocsis** [comment]
Years after, I think the question is still valid and unanswered. Generaly no matter what you do, some of the Code Analysis warning, if they're enabled, are triggered. This problem should be officially solved by Microsoft.

**Reply by huoyaoyuan** [reply]
Talking about the analyzers, they does sometimes introduce false suggestions. They are considered less reliable.

But this feature is definitely a pain in the language.

**Comment by mysteryx93** [comment]
This hasn't been mentioned here; using [Fody.ConfigureAwait](https://github.com/Fody/ConfigureAwait) is an easy solution to that mess.

It would be great to have a built-in solution though.

**Comment by owns** [comment]
ran into this issue today. Had to use 2 lines 😢.
```csharp
var efContext = await _efContextFactory.CreateDbContextAsync(cancellationToken).ConfigureAwait(false);
await using var __ = efContext.ConfigureAwait(false);
```
Cannot use an `out` parameter with async. I feel like this is a false-positive. I don't think `DisposeAsync` should be `ConfigureAwait` since it should be fast enough 🤷‍♂️.

**Reply by mysteryx93** [reply]
If it's "fast enough" you don't need Async at all. That's designed for things that take an undetermined time. Let's say cleaning up temp files, shutting down a hard-drive, doing network stuff, or things like that. Perhaps an operation with 2 seconds timeout. Maybe disposing of a bunch of plugins, each doing their own things.

**Comment by rampaa** [comment]
Just came here to state my annoyance. Using `await using` together with `ConfigureAwait` shouldn't be this inconvenient.

**Comment by RichardR-cg** [comment]
#8778 is a language proposal which describes the various issues surrounding this problem. Please consider upvoting that as a way of representing your annoyance and/or make suggestions on it (because this is a 'question' which has been answered).

**Comment by MattParkerDev** [comment]
The original example can be written as:
```csharp
public async Task FooBar()
{
	await using var test = new AwaitUsingConfigureAwaitTest();
	test.ConfigureAwait(false);
	test.Dummy();
}
```
I don't think this is the worst thing in the world, but I agree that a nicer way would be ideal.

**Reply by 0xfeeddeadbeef** [reply]
That code is not actually awaiting on a configured awaitable. ConfigureAwait method returns a ConfiguredTaskAwaitable struct that you are supposed to await.

**Reply by MattParkerDev** [reply]
Apologies if I'm misunderstanding.

<img width="1346" height="458" alt="image" src="https://github.com/user-attachments/assets/6039f8bf-3397-4310-961f-e6c4a75787a3" />

`await using` is syntactic sugar for, call DisposeAsync when this variable goes out of scope.
In my example, cancellationToken.Register returns a `CancellationTokenRegistration`, which implements `IAsyncDisposable`.
Then, calling `.ConfigureAwait(false)` on the IAsyncDisposable configures how awaits on Tasks returned from DisposeAsync will be called.

I'm not sure I see the issue.

**Reply by 0xfeeddeadbeef** [reply]
While my original comment was wrong about the returned type (it is `ConfiguredAsyncDisposable` instead of `ConfiguredTaskAwaitable`), you still must call DisposeAsync method (via `await using`) of the returned `ConfiguredAsyncDisposable` for it to have any effect.

**Reply by MattParkerDev** [reply]
Thanks for clarifying! To recap for myself (and anyone else reading), these are the only correct ways to call and use ConfigureAwait on an IAsyncDisposable: (Again, please correct me if I'm wrong!)

```csharp
// 'using' statement
// if you don't need access to the CancellationTokenRegistration
await using (cancellationToken.Register(() => buildManager.CancelAllSubmissions()).ConfigureAwait(false))
{
	// disposed outside this scope
}

// 'using' statement
// if you need access to the CancellationTokenRegistration
var test = cancellationToken.Register(() => buildManager.CancelAllSubmissions());
await using (test.ConfigureAwait(false))
{
	// disposed outside this scope
	var cts = test.Token;
};

// 'using' declaration
// if you don't need access to the CancellationTokenRegistration
await using var _ = cancellationToken.Register(() => buildManager.CancelAllSubmissions()).ConfigureAwait(false);

// 'using' declaration
// if you need access to the CancellationTokenRegistration
var test2 = cancellationToken.Register(() => buildManager.CancelAllSubmissions());
await using var __ = test2.ConfigureAwait(false);
var cts2 = test2.Token;
```

Noting that `await using var _` and `await using var __` are not true discards, they are variables, hence the need to have different names. There is a discussion around allowing `await using _ = ...` here: https://github.com/dotnet/csharplang/discussions/8605
